### PR TITLE
loadable_app/wifiapp : Add scenario in binary update sample

### PR DESCRIPTION
**Pull request by YiHyunJin**
Added scenarios to check binary_manager_register_state_changed_callback
and binary_manager_unregister_state_changed_callback functions.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [8fb4d6c2c58836b0d805476ac778fa6454db4b40] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [8fb4d6c2c58836b0d805476ac778fa6454db4b40] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [59342fec44322d4ab9c0d4689e26c919748056fa] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [59342fec44322d4ab9c0d4689e26c919748056fa] - Code Rule Check OK.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 1]
```diff
@@ -70,6 +70,86 @@ static volatile bool inf_flag = true;
...
+	if (read_fd < 0) {
+		fail_cnt++;
+		printf("Failed to open %s: %d, errno: %d\n", binary_info->active_dev, ret, get_errno());
+		return;
+	}
+	write_fd = open(binary_info->inactive_dev, O_WRONLY);
+	if (write_fd < 0) {
+		fail_cnt++;
+		printf("Failed to open %s: %d, errno: %d\n", binary_info->inactive_dev, ret, get_errno());
+		return;
```
Comment (jeongchanKim) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: You'd better close the `read_fd`.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 2]
```diff
@@ -70,6 +70,86 @@ static volatile bool inf_flag = true;
...
+	int read_fd;
+	int write_fd;
+	int ret;
+	int file_size;
+	int read_size;
+	uint32_t crc_hash = 0;
+	uint8_t buffer[BUFFER_SIZE];
+	binary_header_t header_data;
+
+	read_fd = open(binary_info->active_dev, O_RDONLY);
```
Comment (jeongchanKim) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: Please make separate static functions for read and write.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 3]
```diff
@@ -70,6 +70,86 @@ static volatile bool inf_flag = true;
...
+	file_size = header_data.bin_size;
+	while (file_size > 0) {
+		read_size = file_size < BUFFER_SIZE ? file_size : BUFFER_SIZE;
+		ret = read(read_fd, (FAR uint8_t *)buffer, read_size);
+		if (ret != read_size) {
+			printf("Failed to read buffer %s: %d\n", binary_info->active_dev, ret);
+			close(read_fd);
+			return;
+		}
+		ret = write(write_fd, (FAR uint8_t *)buffer, read_size);
```
Comment (jeongchanKim) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: Maybe it takes so much time, please notify to user, now we are in copying.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 4]
```diff
@@ -272,6 +352,11 @@ static void binary_update_run_tests(int repetition_num)
 	printf("\n** Binary Update Example %d-th Iteration.\n", repetition_num);
 
 	binary_update_getinfo(APP_NAME, &pre_bin_info);
+
+	binary_manager_register_state_changed_callback((binmgr_statecb_t *)binary_update_cb, NULL);
+	/* init other partition */
```
Comment (jeongchanKim) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: Please explain more why we want to copy.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 5]
```diff
@@ -70,6 +70,86 @@ static volatile bool inf_flag = true;
 static int fail_cnt = 0;
 static unsigned int new_version = 20190422;
 
+static void binary_update_cb(void)
+{
+	printf(" =================================== \n");
+	printf("          callback funtion           \n");
```
Comment (sunghan-chang) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]:
```
The state changed callback is executed in WIFI. MICOM state is changed.
```

File: loadable_apps/loadable_sample/Kconfig [hunk 6]
```diff
@@ -36,6 +36,13 @@ config EXAMPLES_RECOVERY_TEST
 	---help---
 		Enable the Binary Manager Recovery test
 
+config EXAMPLES_BINARY_UPDATE_TEST
```
Comment (sunghan-chang) [inline loadable_apps/loadable_sample/Kconfig]: Why do you insert this at middle of recovery configs?

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 7]
```diff
@@ -70,6 +70,86 @@ static volatile bool inf_flag = true;
...
+		fail_cnt++;
+		printf("Failed to open %s: %d, errno: %d\n", binary_info->inactive_dev, ret, get_errno());
+		return;
+	}
+	/* read to crc information. */
+	ret = read(read_fd, (FAR uint8_t *)&crc_hash, sizeof(uint32_t));
+	if (ret == ERROR) {
+		printf("Failed to read crc %s\n", binary_info->active_dev);
+		close(read_fd);
+		return;
```
Comment (sunghan-chang) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: need to close write_fd as well

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 8]
```diff
@@ -70,6 +70,86 @@ static volatile bool inf_flag = true;
...
+	}
+
+	/* Write to crc information. */
+	ret = write(write_fd, (FAR uint8_t *)&crc_hash, sizeof(uint32_t));
+	if (ret == ERROR) {
+		printf("Failed to write crc %s\n", binary_info->inactive_dev);
+		close(write_fd);
+		return;
```
Comment (sunghan-chang) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: close read_fd

Comment (sunghan-chang) [reply loadable_apps/loadable_sample/wifiapp/binary_update.c]: Please check it in entire this file.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c (hunk 3 above)
Comment (YiHyunJin) [reply loadable_apps/loadable_sample/wifiapp/binary_update.c]: Done

File: loadable_apps/loadable_sample/wifiapp/binary_update.c (hunk 7 above)
Comment (YiHyunJin) [reply loadable_apps/loadable_sample/wifiapp/binary_update.c]: Done.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c (hunk 2 above)
Comment (YiHyunJin) [reply loadable_apps/loadable_sample/wifiapp/binary_update.c]: Thanks. I changed the location.

**Comment by seinfra** [comment]
Target : [5b00a4ac22fb39a07f0b8da99320cd8e394a93d1] - Code Rule Check Result:
loadable_apps/loadable_sample/wifiapp/binary_update.c:76: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxV)

**Comment by seinfra** [comment]
Target : [5b00a4ac22fb39a07f0b8da99320cd8e394a93d1] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [5c16bacea5b69d6d007470257192cff83a42fe88] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [5c16bacea5b69d6d007470257192cff83a42fe88] - Code Rule Check OK.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 9]
```diff
@@ -263,15 +356,46 @@ static void binary_update_reload(char *name)
 	}
 }
 
+static void binary_update_register_state_changed_callback()
```
Comment (sunghan-chang) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: (void)

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 10]
```diff
@@ -263,15 +356,46 @@ static void binary_update_reload(char *name)
...
+	printf("\n** Binary Update Register state changed callback test.\n");
+	ret = binary_manager_register_state_changed_callback((binmgr_statecb_t *)binary_update_cb, NULL);
+	if (ret == OK) {
+		printf("Register state changed callback SUCCESS\n");
+	} else {
+		printf("Register state changed callback FAIL, ret %d\n", ret);
+	}
+}
+
+static void binary_update_unregister_state_changed_callback()
```
Comment (sunghan-chang) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: same as previous

**Comment by seinfra** [comment]
Target : [2d1e391dc57aae444ee2ac69b72c2d5f7ac7cf78] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [2d1e391dc57aae444ee2ac69b72c2d5f7ac7cf78] - Code Rule Check OK.

File: loadable_apps/loadable_sample/wifiapp/binary_update.c [hunk 11]
```diff
@@ -68,8 +68,102 @@ typedef struct binary_header_s binary_header_t;
...
+static void binary_update_copy_binary(binary_update_info_t *binary_info)
+{
+	int read_fd;
+	int write_fd;
+	int ret;
+	int total_size;
+	int copy_size;
+	int read_size;
+	uint32_t crc_hash = 0;
+	uint8_t buffer[BUFFER_SIZE];
```
Comment (jeongchanKim) [inline loadable_apps/loadable_sample/wifiapp/binary_update.c]: Please add `#include <stdint.h>`

File: loadable_apps/loadable_sample/wifiapp/wifiapp_internal.h [hunk 12]
```diff
@@ -25,4 +25,5 @@ void recovery_test(void);
 void messaging_test(void);
 #endif
 
+void binary_update_test(void);
```
Comment (jeongchanKim) [inline loadable_apps/loadable_sample/wifiapp/wifiapp_internal.h]: Please add `#ifdef CONFIG_EXAMPLES_BINARY_UPDATE_TEST`

**Comment by seinfra** [comment]
Target : [33504548c0171e1819e72509a98806fdca5b0650] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [33504548c0171e1819e72509a98806fdca5b0650] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e4fd86a2668eb41b17d41f2f37947e0df72598f7] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e4fd86a2668eb41b17d41f2f37947e0df72598f7] - Code Rule Check (C++) OK.

**Pull request merged**